### PR TITLE
Review/metadata persistence

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: src/java/org/apache/cassandra/service/accord/async/AsyncWriter.java [hunk 1]
```diff
@@ -211,7 +212,8 @@ private void denormalize(AccordCommand command, AsyncContext context)
         }
 
         Map<PartitionKey, AccordCommandsForKey> addCfkToCtx = new HashMap<>();
-        if (CommandSummaries.commandsPerKey.needsUpdate(command))
+        // NotWitnessed has 0 state, and doesn't know about the tx keys, so unable to populate summary
+        if (command.status() != Status.NotWitnessed && CommandSummaries.commandsPerKey.needsUpdate(command))
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/async/AsyncWriter.java]: not actually sure this is the correct thing to do... a command may be dependent on a `NotWitnessed`, but not sure if this change is the proper change...

File: src/java/org/apache/cassandra/service/accord/async/AsyncWriter.java [hunk 2]
```diff
@@ -227,7 +229,8 @@ private void denormalize(AccordCommand command, AsyncContext context)
 
     private void denormalize(AsyncContext context)
     {
-        context.commands.items.values().forEach(command -> denormalize(command, context));
+        // need to clone "values" as denormalize will mutate it
+        new ArrayList<>(context.commands.items.values()).forEach(command -> denormalize(command, context));
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/async/AsyncWriter.java:233]: this fixes the concurrent modification issue reported

File: test/distributed/org/apache/cassandra/distributed/impl/Instance.java [hunk 3]
```diff
@@ -44,6 +44,9 @@
 
 import com.google.common.annotations.VisibleForTesting;
 
+import org.slf4j.Logger;
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/impl/Instance.java:47]: this change was to make sure I was dropping the correct messages.  I think trunk got a similar version already

File: test/distributed/org/apache/cassandra/distributed/test/accord/AccordIntegrationTest.java [hunk 4]
```diff
@@ -62,8 +66,12 @@
 import static org.apache.cassandra.service.accord.db.AccordUpdate.UpdatePredicate.Type.EQUAL;
 import static org.apache.cassandra.service.accord.db.AccordUpdate.UpdatePredicate.Type.NOT_EXISTS;
 
+//TODO there are too many new clusters, this will cause Metaspace issues.  Once Schema and topology are integrated, can switch
+// to a shared cluster with isolated tables
 public class AccordIntegrationTest extends TestBaseImpl
 {
+    private static final String keyspace = "ks";
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/accord/AccordIntegrationTest.java:73]: no need to have `+ System.currentTimeMillis();` as we isolate at the cluster level.  Also, due to the fun of class loaders, having a static field that uses `+ System.currentTimeMillis();` will have different values depending on the CL...  so simpler to have a constant value

File: test/distributed/org/apache/cassandra/distributed/test/accord/AccordIntegrationTest.java [hunk 5]
```diff
@@ -72,17 +80,29 @@ private static void assertRow(Cluster cluster, String query, int k, int c, int v
...
             cluster.forEach(node -> node.runOnInstance(() -> AccordService.instance.setCacheSize(0)));
 
+            fn.accept(cluster);
+        }
+    }
+
+    private static Cluster createCluster() throws IOException
+    {
+        // need to up the timeout else tests get flaky
+        return init(Cluster.build(2).withConfig(c -> c.set("write_request_timeout_in_ms", TimeUnit.SECONDS.toMillis(10))).start());
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/accord/AccordIntegrationTest.java:99]: I found tests were flaky due to timeouts, so bumping to 10s made them more stable

File: test/distributed/org/apache/cassandra/distributed/test/accord/AccordIntegrationTest.java [hunk 6]
```diff
@@ -166,29 +170,30 @@ public void testRecovery() throws Throwable
 
             awaitAsyncApply(cluster);
 
-            assertRow(cluster, "SELECT * FROM " + keyspace + ".tbl WHERE k=0 AND c=0", 0, 0, 1);
+            //TODO why is this flakey?  Also why does .close hang on Accord?
+            assertRow(cluster, "SELECT * FROM " + keyspace + ".tbl WHERE k=0 AND c=0", 0, 0, 2);
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/accord/AccordIntegrationTest.java:174]: I hope this is 100% is due to the fact `awaitAsyncApply` does *not* know when accord tx is complete... seeing `0, 0, 1` and `0, 0, 2`... 10s might not be enough anymore due to timeouts being 10s as well?

File: test/distributed/org/apache/cassandra/distributed/impl/Instance.java [hunk 7]
```diff
@@ -743,7 +755,8 @@ public Future<Void> shutdown(boolean graceful)
                                 () -> SSTableReader.shutdownBlocking(1L, MINUTES),
                                 () -> shutdownAndWait(Collections.singletonList(ActiveRepairService.repairCommandExecutor())),
                                 () -> ScheduledExecutors.shutdownNowAndWait(1L, MINUTES),
-                                () -> SnapshotManager.shutdownAndWait(1L, MINUTES)
+                                () -> SnapshotManager.shutdownAndWait(1L, MINUTES),
+                                () -> AccordService.instance.shutdownAndWait(1l, MINUTES)
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/impl/Instance.java:759]: cleanup accord memory

File: src/java/org/apache/cassandra/service/accord/AccordService.java [hunk 8]
```diff
@@ -59,6 +64,7 @@ private AccordService()
                              new AccordAgent(),
                              scheduler,
                              AccordCommandStores::new);
+        this.nodeShutdown = toShutdownable(node);
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/AccordService.java:67]: don't like this, but will send a PR for accord to cleanup a few things (shutdown able, gradle changes, etc.)

**Review by bdeggleston — APPROVED** [review]
looks good, with one very minor tweak

**Pull request merged**